### PR TITLE
genmsg-extras: Added a BASE_DIR argument to add_service_files to be able...

### PR DESCRIPTION
**Pull request by dreuter**
... to compile autogenerated messages.

### CODE REVIEW

**Comment by dirk-thomas** [comment]
Please describe your use case for this patch. It also looks like that the argument is very similar to the existing argument `DIRECTORY`.

**Comment by dreuter** [comment]
Okay, I'm sorry that I didn't provide them before.

It is quite similar to `DIRECTORY` but it does not take a relative path, but an absolute. For add_message_files both arguments exist.

A typical usecase would be autogenerated service files. I am currently writing a database API, which uses ROS services. Since I don't want to generate a service for each message used in our system, I wanted to create an autogenerator.

This could be used like this:

``` CMake
autogenerate_srvs(autogen_srvs a b c d)

add_service_files(
  FILES ${autogen_srvs}
  BASE_DIR ${CMAKE_CURRENT_BINARY_DIR}
)
```

where autogenerate_srvs would write some services (for each argument) in `${CMAKE_CURRENT_BINARY_DIR}`

There are some (hacky) ways around this, but I think it is a good idea to integrate it, especially because `add_message_files` has such an argument.

**Comment by dirk-thomas** [comment]
The `BASE_DIR` argument for `add_message_files` solely exists for backward compatibility with actions.

I don't think we should mimic the same (actually pretty weird) arguments to `add_service_files`. Instead the `DIRECTORY` argument could be modified that it accepts an absolute path. Can you please update your pull request that way?

**Comment by dreuter** [comment]
I actually tried `DIRECTORY` first and I thought it would handle the path as absolute, if I prepend a slash. So I will give it a try :)

Is `BASE_DIR` used somewhere? I'm just asking because I am wondering, if I should remove it then.

**Comment by dreuter** [comment]
I pushed the new changes. DIRECTORY now handles absolute and relative paths.

I also added a commit, which removes the BASE_DIR argument, so you can omit it, if necessary.

**Comment by dreuter** [comment]
I forgot to mention, that deleting BASE_DIR will break action files at the moment. I will try to fix this.

I also created a repository, which tests the different calling methods: https://github.com/dreuter/test_genmsg

**Comment by dreuter** [comment]
My pull request in common_msgs should fix the problem mentioned above.

See: https://github.com/ros/common_msgs/pull/43

File: cmake/genmsg-extras.cmake.em [hunk 1]
```diff
@@ -74,7 +74,7 @@ macro(_prepend_path ARG_PATH ARG_FILES ARG_OUTPUT_VAR)
 endmacro()
 
 macro(add_message_files)
-  cmake_parse_arguments(ARG "NOINSTALL" "DIRECTORY;BASE_DIR" "FILES" ${ARGN})
+  cmake_parse_arguments(ARG "NOINSTALL" "DIRECTORY" "FILES" ${ARGN})
```
Comment (dirk-thomas) [inline cmake/genmsg-extras.cmake.em]: Removing public API from an already release distro is not feasible.

**Comment by dreuter** [comment]
I fixed this. The public API now remains unchanged... Could you please merge it?

**Comment by dirk-thomas** [comment]
Thank you for the patch - I committed a modified version which is a bit more concise: 80c0d2a0ee8c618f775d3bfed8d31028e142f2d8

**Pull request closed without merging**